### PR TITLE
systemctl: Add --with-dependencies switch

### PR DESCRIPTION
**Pull request by kkuehlz**

### CODE REVIEW

File: src/systemctl/systemctl.c [hunk 1]
```diff
@@ -799,6 +800,107 @@ static int expand_names(sd_bus *bus, char **names, const char* suffix, char ***r
...
+                },
+                [DEPENDENCY_AFTER] = {
+                        { "After",       "as", NULL, offsetof(struct DependencyStatusInfo, dep[0]) },
+                        {}
+                },
+                [DEPENDENCY_BEFORE] = {
+                        { "Before",      "as", NULL, offsetof(struct DependencyStatusInfo, dep[0]) },
+                        {}
+                },
+        };
```
Comment (poettering) [inline src/systemctl/systemctl.c:831]: looking at this, why is this even so complex? I mean, we'll merge the 5 strv arrays anyway later on, so we can just specify the same array here in instead of dep[0], dep[1], … The default bus property parsers append anyway, so no need to complicate stuff

(i understand this is older code you just moved, but maybe we should fix that right away, too)

File: src/systemctl/systemctl.c [hunk 2]
```diff
@@ -799,6 +800,107 @@ static int expand_names(sd_bus *bus, char **names, const char* suffix, char ***r
...
+        assert(bus);
+        assert(ret);
+
+        STRV_FOREACH(name, names) {
+                _cleanup_strv_free_ char **deps = NULL;
+                _cleanup_free_ char *n = NULL;
+                char **dep;
+
+                n = strdup(*name);
+                if (strv_consume(&with_deps, TAKE_PTR(n)) < 0)
```
Comment (poettering) [inline src/systemctl/systemctl.c]: strv_extend()

File: src/systemctl/systemctl.c [hunk 3]
```diff
@@ -799,6 +800,107 @@ static int expand_names(sd_bus *bus, char **names, const char* suffix, char ***r
...
+                        return log_oom();
+
+                list_dependencies_get_dependencies(bus, *name, &deps);
+
+                STRV_FOREACH(dep, deps)
+                        if (strv_consume(&with_deps, TAKE_PTR(*dep)) < 0)
+                                return log_oom();
```
Comment (poettering) [inline src/systemctl/systemctl.c]: if this fails in the middle, then the beginning will be nulled out, and the end will not be, but strv_free() (as called due to _cleanup_strv_free() will then stop freeing in the beginning...

**Comment by keur** [comment]
@poettering changes made. i changed tried changing deps to one array, but it wasn't always appending and would set the array to NULL.

File: src/systemctl/systemctl.c [hunk 4]
```diff
@@ -799,6 +800,91 @@ static int expand_names(sd_bus *bus, char **names, const char* suffix, char ***r
...
+                        {}
+                },
+                [DEPENDENCY_BEFORE] = {
+                        { "Before",      "as", NULL, offsetof(struct DependencyStatusInfo, deps) },
+                        {}
+                },
+        };
+
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_strv_free_ char **ret = NULL;
```
Comment (mrc0mmand) [inline src/systemctl/systemctl.c]:
```
    archlinux_systemd_ci: ../src/systemctl/systemctl.c:836:36: error: unused variable 'ret' [-Werror,-Wunused-variable]
    archlinux_systemd_ci:         _cleanup_strv_free_ char **ret = NULL;
    archlinux_systemd_ci:                                    ^
    archlinux_systemd_ci: 1 error generated.
```

Comment (kkuehlz) [reply src/systemctl/systemctl.c]: [Oops](https://i.fluffy.cc/TH6LWWcp0CwpXN9jzCDKcbklmX4gx4fL.jpeg). Thanks!

File: src/shared/bus-util.c [hunk 5]
```diff
@@ -1127,7 +1127,7 @@ static int map_basic(sd_bus *bus, const char *member, sd_bus_message *m, unsigne
                 if (r < 0)
```
Comment (keszybz) [inline src/shared/bus-util.c:1127]: Please add an explanation to the commit message why this change is appropriate, including the effect on the callers. This is definitely not a trivial change!

File: src/systemctl/systemctl.c (hunk 1 above)
Comment (keszybz) [reply src/systemctl/systemctl.c:831]: This still needs trimming. Why is `info` a structure?

Also, this is missing error handling (the code was already borked before...). `bus_map_all_properties` will modify the output argument even if it fails halfway. So the dep argument should be something like `_cleanup_strv_free_ char **deps = NULL;`.

**Comment by keur** [comment]
@keszybz Changes made

**Comment by poettering** [comment]
hmm, we try to merge "perfect" PRs, i.e. where each commit stands on its own feet, and is "perfect" and not corrected in later commits of the same PR, so that things remain perfectly bisectable. 

In this PR the first commit is majorly reworked by the last commit. Could you squash that so that the PR only contains logically separate patches, but not stuff reworking itself? thanks!

**Comment by mrc0mmand** [comment]
@keur I suspect it was a build fail in Travis/Azure (as someone has already sneakily rescheduled it! :-)), which was caused by #14520, so it was definitely not your fault.

**Comment by keur** [comment]
Thanks @mrc0mmand !

Changes made cc @poettering @keszybz

File: src/systemctl/systemctl.c [hunk 6]
```diff
@@ -799,6 +800,88 @@ static int expand_names(sd_bus *bus, char **names, const char* suffix, char ***r
...
+        assert(bus);
+        assert(ret);
+
+        STRV_FOREACH(name, names) {
+                _cleanup_strv_free_ char **deps = NULL;
+
+                if (strv_extend(&with_deps, *name) < 0)
+                        return log_oom();
+
+                list_dependencies_get_dependencies(bus, *name, &deps);
```
Comment (poettering) [inline src/systemctl/systemctl.c]: if we ignore errors (which is OK in many cases), then let's make sure to cast the result to (void), to indicate that we do so knowingly and to give static checkers a hint

File: src/systemctl/systemctl.c [hunk 7]
```diff
@@ -5959,6 +6004,18 @@ static int cat(int argc, char *argv[], void *userdata) {
...
+
+                r = append_unit_dependencies(bus, names, &names_with_deps);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to append unit dependencies: %m");
+
+                strv_free(names);
+                names = TAKE_PTR(names_with_deps);
+        }
+
+
```
Comment (poettering) [inline src/systemctl/systemctl.c:6022]: spurious double newline

File: src/systemctl/systemctl.c [hunk 8]
```diff
@@ -5959,6 +6004,18 @@ static int cat(int argc, char *argv[], void *userdata) {
...
+        if (arg_with_dependencies) {
+                _cleanup_strv_free_ char **names_with_deps = NULL;
+
+                r = append_unit_dependencies(bus, names, &names_with_deps);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to append unit dependencies: %m");
+
+                strv_free(names);
+                names = TAKE_PTR(names_with_deps);
+        }
```
Comment (poettering) [inline src/systemctl/systemctl.c]: this code you add in a pretty similar way a bit earlier. Maybe convert this into a function of its own:

```c
static int maybe_extend_with_unit_dependencies(sd_bus *bus, char ***list) { 
…
```

and that function would leave the specified list untouched if arg_with_dependencies is false, and otherwise augment it with the dependencies, updating the specified argument in place.

File: src/systemctl/systemctl.c [hunk 9]
```diff
@@ -7945,6 +8002,8 @@ static int systemctl_help(void) {
                "  -l --full              Don't ellipsize unit names on output\n"
                "  -r --recursive         Show unit list of host and local containers\n"
                "     --reverse           Show reverse dependencies with 'list-dependencies'\n"
+               "     --with-dependencies Show unit dependencies with 'status' 'cat',\n"
```
Comment (poettering) [inline src/systemctl/systemctl.c]: a comma is missing between status and cat?

**Comment by keur** [comment]
@poettering round 3

**Comment by poettering** [comment]
Excellent! Thank you!
